LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClientConnectionManager.cs

```diff
@@ -1,162 +1,399 @@
+                    };
+
+                    this.exclusiveProcessor.Processed += (_, args) =>
+                    {
+                        var ((id, name), outcome) = args;
+                        someLogger.LogDebug(@"Invoked ""{Name}"" ({Id}); status = {Status}, run-time = {RunTime}, wait-time = {WaitTime}",
+                                            name, id, outcome.Task.Status, outcome.RunDuration, outcome.WaitDuration);
+                    };
+
+                    this.exclusiveProcessor.Interrupted += (_, p) =>
+                    {
+                        var (interrupted, interrupting) = p;
+                        someLogger.LogDebug(@"Interrupted ""{Name}"" ({Id}) by {InterruptingName} ({InterruptingId})",
+                                            interrupted.Name, interrupted.Id,
+                                            interrupting.Name, interrupting.Id);
+                    };
+                }
             }
-        }
 
-        private static string GetConnectionCacheKey(DevEui devEui) => string.Concat("connection:", devEui);
+            public DevEui DevEui => this.device.DevEUI;
+            public TimeSpan KeepAliveTimeout => TimeSpan.FromSeconds(this.device.KeepAliveTimeout);
 
-        private static string GetScheduleCacheKey(DevEui devEui) => string.Concat("connection:schedule:", devEui);
+            public void Dispose() => this.client?.Dispose();
 
-        public ILoRaDeviceClient GetClient(LoRaDevice loRaDevice)
-        {
-            if (loRaDevice is null) throw new ArgumentNullException(nameof(loRaDevice));
+            public event EventHandler? EnsureConnectedSucceeded;
 
-            return this.managedConnections.TryGetValue(GetConnectionCacheKey(loRaDevice.DevEUI), out var managedConnection)
-                 ? managedConnection.DeviceClient
-                 : throw new ManagedConnectionException($"Connection for device {loRaDevice.DevEUI} was not found");
-        }
+            private Task<T> InvokeExclusivelyAsync<T>(Func<ILoRaDeviceClient, Task<T>> processor,
+                                                      [CallerMemberName] string? callerName = null) =>
+                InvokeExclusivelyAsync(doesNotRequireOpenConnection: false, processor, callerName);
 
-        public void Register(LoRaDevice loRaDevice, ILoRaDeviceClient loraDeviceClient)
-        {
-            if (loRaDevice is null) throw new ArgumentNullException(nameof(loRaDevice));
+            private async Task<T> InvokeExclusivelyAsync<T>(bool doesNotRequireOpenConnection,
+                                                            Func<ILoRaDeviceClient, Task<T>> processor,
+                                                            [CallerMemberName] string? callerName = null)
+            {
+                _ = Interlocked.Increment(ref this.pid);
+                return await this.exclusiveProcessor.ProcessAsync(new Process(this.pid, callerName!), async () =>
```

Comment:
where do we need that pid? couldn't figure it out.

Reply:
I saw we print the "id" in the callbacks. Not sure what we will use this for though?

Reply:
For now, this is purely for diagnostics purpose to understand the sequence in which operations were invoked. If you think it's not providing any value and just adding noise to logs then I can remove it.

Reply:
I think you may also need it to ensure uniqueness of the operation? But I might be wrong. It's ok to have that sequence number, however, you may want to rename it so it is obvious, that it is a sequence number for the operations?

Reply:
Renamed to `operationSequenceNumber` with 5a3e07d0391031bfd8bf0e1a9f782faa78f675ca.